### PR TITLE
Fix OBJECT_ID() to exclude indexes

### PR DESCRIPTION
**Pull request by roshan0708**
### Description

This commit modifies `OBJECT_ID` function to exclude indexes when searching for objects including temp table as well, ensuring that it correctly returns OIDs for constraints and other objects without mistakenly returning index OIDs.

### Issues Resolved

JIRA-5303

### Test Scenarios Covered ###
* **Use case based -** y

* **Boundary conditions -**

* **Arbitrary inputs -**

* **Negative test cases -**

* **Minor version upgrade tests -**

* **Major version upgrade tests -**

* **Performance tests -**

* **Tooling impact -**

* **Client tests -**

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: contrib/babelfishpg_tsql/runtime/functions.c [hunk 1]
```diff
@@ -2361,13 +2361,39 @@ object_id(PG_FUNCTION_ARGS)
 			} 
 			else if (enr == NULL)
 			{
-				result = get_relname_relid((const char *) object_name, LookupNamespaceNoError("pg_temp"));
+				Oid temp_nsp_oid = LookupNamespaceNoError("pg_temp");
+				if (OidIsValid(temp_nsp_oid))
+				{
+					result = get_relname_relid((const char *) object_name, temp_nsp_oid);
+				}
```
Comment (thephantomthief) [inline contrib/babelfishpg_tsql/runtime/functions.c]: Is this really necessary? I'm assuming even before if pg_temp lookup fails we would still set result as `InvalidOid`

File: contrib/babelfishpg_tsql/runtime/functions.c [hunk 2]
```diff
@@ -2361,13 +2361,39 @@ object_id(PG_FUNCTION_ARGS)
...
+					{
+						result = relid;
+					}
+				}
+
+				ReleaseSysCache(class_tuple);
+			}
 
 			if (OidIsValid(relid) && pg_class_aclcheck(relid, user_id, ACL_SELECT) == ACLCHECK_OK)
 			{
```
Comment (thephantomthief) [inline contrib/babelfishpg_tsql/runtime/functions.c:2392]: Should we remove this check then?

File: contrib/babelfishpg_tsql/runtime/functions.c [hunk 3]
```diff
@@ -2361,13 +2361,39 @@ object_id(PG_FUNCTION_ARGS)
...
+				/* Check if the relation is not an index */
+				if (classform->relkind != RELKIND_INDEX)
+				{
+					relid = classform->oid;
+
+					/* Perform ACL check */
+					if (pg_class_aclcheck(relid, user_id, ACL_SELECT) == ACLCHECK_OK)
+					{
+						result = relid;
+					}
```
Comment (thephantomthief) [inline contrib/babelfishpg_tsql/runtime/functions.c]: we should also check for `OidIsValid(relid)`

File: contrib/babelfishpg_tsql/runtime/functions.c (hunk 1 above)
Comment (roshan0708) [reply contrib/babelfishpg_tsql/runtime/functions.c]: You're right it's not strictly needed, but I think it adds clarity and safety. Unless it slows things down, I'd like to keep it.

File: contrib/babelfishpg_tsql/runtime/functions.c (hunk 2 above)
Comment (roshan0708) [reply contrib/babelfishpg_tsql/runtime/functions.c:2392]: Good catch! I missed removing this. Thanks for spotting that.

**Review by thephantomthief — APPROVED** [review]

File: contrib/babelfishpg_tsql/runtime/functions.c [hunk 4]
```diff
@@ -2361,17 +2361,38 @@ object_id(PG_FUNCTION_ARGS)
...
 		{
-			/* search in pg_class by name and schema oid */
-			Oid			relid = get_relname_relid((const char *) object_name, schema_oid);
+			HeapTuple	class_tuple;
+			Oid			relid = InvalidOid;
+
+			/* Perform a single lookup in pg_class by name and schema oid */
+			class_tuple = SearchSysCache2(RELNAMENSP, CStringGetDatum(object_name), ObjectIdGetDatum(schema_oid));
 
-			if (OidIsValid(relid) && pg_class_aclcheck(relid, user_id, ACL_SELECT) == ACLCHECK_OK)
```
Comment (tanscorpio7) [inline contrib/babelfishpg_tsql/runtime/functions.c:2379]: We can just add another condition `get_rel_relkind(relid) != RELKIND_INDEX`

File: contrib/babelfishpg_tsql/runtime/functions.c [hunk 5]
```diff
@@ -2361,17 +2361,38 @@ object_id(PG_FUNCTION_ARGS)
 			} 
 			else if (enr == NULL)
 			{
-				result = get_relname_relid((const char *) object_name, LookupNamespaceNoError("pg_temp"));
+				Oid temp_nsp_oid = LookupNamespaceNoError("pg_temp");
+				if (OidIsValid(temp_nsp_oid))
+				{
+					result = get_relname_relid((const char *) object_name, temp_nsp_oid);
```
Comment (tanscorpio7) [inline contrib/babelfishpg_tsql/runtime/functions.c]: `get_relname_relid` is used in multiple other places in OBJECT ID. Do we need to add relkind check at all the places ?

**Review by tanscorpio7 — CHANGES_REQUESTED** [review]

File: contrib/babelfishpg_tsql/runtime/functions.c (hunk 5 above)
Comment (roshan0708) [reply contrib/babelfishpg_tsql/runtime/functions.c]: We don't need to check this at all the places since only constraints can be confused as indexes.

Comment (tanscorpio7) [reply contrib/babelfishpg_tsql/runtime/functions.c]: What about cases like `SELECT OBJECT_ID('some constraint name', 's')`

**Review by tanscorpio7 — APPROVED** [review]

**Review by thephantomthief — APPROVED** [review]

**Review by shardgupta — APPROVED** [review]
